validator_client/src/beacon_node_fallback.rs

```diff
@@ -0,0 +1,424 @@
+    pub async fn num_total(&self) -> usize {
+        self.candidates.len()
+    }
+
+    /// The count of synced and ready candidates.
+    pub async fn num_synced(&self) -> usize {
+        let mut n = 0;
+        for candidate in &self.candidates {
+            if candidate.status(RequireSynced::Yes).await.is_ok() {
+                n += 1
+            }
+        }
+        n
+    }
+
+    /// The count of candidates that are online and compatible, but not necessarily synced.
+    pub async fn num_available(&self) -> usize {
+        let mut n = 0;
+        for candidate in &self.candidates {
+            if candidate.status(RequireSynced::No).await.is_ok() {
+                n += 1
+            }
+        }
+        n
+    }
+
+    /// Loop through any `self.candidates` that we don't think are online, compatible or synced and
+    /// poll them to see if their status has changed.
+    ///
+    /// We do not poll nodes that are synced to avoid sending additional requests when everything is
+    /// going smoothly.
+    pub async fn update_unready_candidates(&self) {
+        for candidate in &self.candidates {
+            // There is a potential race condition between having the read lock and the write
+            // lock. The worst case of this race is running `try_become_ready` twice, which is
+            // acceptable.
+            //
+            // Note: `RequireSynced` is always set to false here. This forces us to recheck the sync
+            // status of nodes that were previously not-synced.
+            if candidate.status(RequireSynced::Yes).await.is_err() {
+                // There exists a race-condition that could result in `refresh_status` being called
+                // when the status does not require refreshing anymore. This deemed is an
+                // acceptable inefficiency.
+                let _ = candidate
+                    .refresh_status(self.slot_clock.as_ref(), &self.spec, &self.log)
+                    .await;
+            }
+        }
+    }
+
+    /// Run `func` against each candidate in `self`, returning immediately if a result is found.
+    /// Otherwise, return all the errors encountered along the way.
+    ///
+    /// First this function will try all nodes with a suitable status. If no candidates are suitable
+    /// or all the requests fail, it will try updating the status of all unsuitable nodes and
+    /// re-running `func` again.
+    pub async fn first_success<'a, F, O, Err, R>(
+        &'a self,
+        require_synced: RequireSynced,
+        func: F,
+    ) -> Result<O, AllErrored<Err>>
+    where
+        F: Fn(&'a BeaconNodeHttpClient) -> R,
+        R: Future<Output = Result<O, Err>>,
+    {
+        let mut errors = vec![];
+        let mut to_retry = vec![];
+
+        // First pass: try `func` on all ready candidates.
+        for candidate in &self.candidates {
+            if let Err(e) = candidate.status(require_synced).await {
```

Comment:
Good catch, this is a useful feature!

I think that https://github.com/blacktemplar/lighthouse/commit/b0bebafc3bc86e1c777960cabc9ab42cf8e93d37 also has some side effects, though. Consider this scenario:

- The VC has 2x BN; BN1 & BN2.
- BN1 fails after a timeout.
- BN2 not synced.

Now, we make a request that has `RequireSynced::No`. The following will happen:

- Both BNs will fail on the first check.
- The second check for BN1 will fail with a timeout
- The second check for BN2 will succeed.

In this scenario we've added an unnecessary cost of one timeout to this call.

I think I've managed to avoid the extra cost in f6e148620, let me know what you think :)

Reply:
Hey Paul, in your current solution I first didn't understand why rerunning all online candidates in the second pass and I wrote this: https://github.com/sigp/lighthouse/commit/e469e85e0c32d22e6a29fd3757fd0b2c48a5894a

but then I understood that if in the first run all candidates failed then there are no "online" candidates and all candidates are either unsynced or otherwise unavailable. So I think your solution is good. What we should consider (and maybe mention in a comment) what happens if the online status changes for some candidates between the first and the second pass (due to asynchronicity). Then we might rerun in the second pass the `func` for some candidates. Is that wanted behavior or not (if not we could use the changes I proposed in https://github.com/sigp/lighthouse/commit/e469e85e0c32d22e6a29fd3757fd0b2c48a5894a).

Otherwise, I am now very happy with the current solution :).

Unfortunately, we a fmt error (might be from my cherry picked changes, I am so sorry) and I can't push to your branch.

Reply:
> Just saw that in your current solution we would add offline candidates twice to to_retry and therefore check them twice in the third pass. 

Good catch! Thank you :)

> Maybe my proposed changes in e469e85 are not so bad after all (I think they make the behavior also more explicit).

I agree, it is more explicit and I like it. I've cherry-picked that comment. However, I did also add d17d36a12 since your changes meant that we don't try to re-check the nodes sync-status during the third loop. Let me know if you see any problems with this.

I can't believe how fiddly this task is! :sweat_smile:

Reply:
I thought rechecking the status is not necessary for unsynced nodes, since they get executed in the second pass anyway, regardless what their (real) status is. So if we get to the third pass that means the unsynced nodes all errored (maybe the calling function errors if the bn is too far behind), then I am not sure if it is useful to do the same call again if the node is still unsynced.

So basically if we allow unsynced nodes this means for me (after trying all synced nodes) unsynced is an accepted status. Therefore we don't recheck the status for them (we also don't recheck the status of all synced nodes if they error the first time).

Reply:
> I thought rechecking the status is not necessary for unsynced nodes, since they get executed in the second pass anyway, regardless what their (real) status is.

Oh yeah, you're right. I missed this! I'll remove that extra line I added.

> Is it correct behavior to always assume the status is offline if the function errors? As described above it could happen that an unsynced node gets set to offline because the function errors if a node is too far behind. This would result in the next call (with allow_unsynced == true) that the node is considered offline and therefore only checked in the third pass. But maybe that is acceptable in this scenario?

Yeah, it isn't ideal that we just set them to offline regardless of the failure type. The alternative to this is trying to match on the `reqwest` errors, but I think we might end up opening a can of worms there. It's not ideal in the scenario you described, but it eventually works out in the end so I think we can deem it acceptable. Let me know if you disagree.

> PS: Can you merge in the unstable branch to let CI run again :)?

Done! Correct me if I'm wrong, but I think we'd be in a good place to get this merged, once CI passes. Yay!